### PR TITLE
Add stop-watch class (with Python bindings) for profiling code.

### PR DESCRIPTION
**Pull request by andreadelprete**
I've added a stop-watch class which is very useful for doing accurate profiling of the code. This is important for the work that I plan to do to optimize the performance of Crocoddyl with high-order integration schemes. Profiling is off by default so that it does not affect computation times.

### CODE REVIEW

**Comment by proyan** [comment]
Nice! Thanks for the work Andrea!

File: include/crocoddyl/core/utils/stop-watch.hpp
Comment (jcarpent) [inline include/crocoddyl/core/utils/stop-watch.hpp:177]: This code is missing the notion of const-ref.

**Review by jcarpent — COMMENTED** [review]
Small about timings of code. getTimeOfDay and other related methods such as std::chrono may impact the real effective timings. In other words, they slow down the process and measured timings should be considered rather as upper-bounds.

File: include/crocoddyl/core/utils/stop-watch.hpp
Comment (cmastalli) [inline include/crocoddyl/core/utils/stop-watch.hpp:164]: Could you modify all the functions comments to be doxygen compatible?

File: src/core/utils/stop-watch.cpp
Comment (cmastalli) [inline src/core/utils/stop-watch.cpp:6]: Same here

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [inline src/core/solvers/ddp.cpp:360]: We have two solveInPlace functions. It would be nice to include both like: 'Quuinv_Qux'

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [inline src/core/solvers/ddp.cpp:347]: You also need to include this code for BoxDDP::computeGains

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [inline src/core/solvers/ddp.cpp:128]: It would be nice to include the BoxDDP solver

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [inline src/core/solvers/ddp.cpp:300]: Also BoxDDP solver here

**Review by cmastalli — COMMENTED** [review]
Thanks @andreadelprete for this nice work.

The most relevant feedback is about propagating the profiling code to the missing pieces in FDDP, BoxDDP and BoxFDDP solvers. There are also other places that might be useful (e.g., shooting problem functions).

File: include/crocoddyl/core/utils/Stdafx.hh
Comment (andreadelprete) [reply include/crocoddyl/core/utils/Stdafx.hh:6]: Unfortunately there is no license specified.

File: src/core/solvers/ddp.cpp
Comment (andreadelprete) [reply src/core/solvers/ddp.cpp:128]: If the BoxDDP solver inherits and uses this method, then it will be profiled. However, it will appear with the name "SolverDDP::computeDirection". The only way around this is to overload computeDirection in the BoxDDP solver

File: src/core/solvers/ddp.cpp
Comment (andreadelprete) [reply src/core/solvers/ddp.cpp:300]: Same comment as above

File: src/core/solvers/ddp.cpp
Comment (andreadelprete) [reply src/core/solvers/ddp.cpp:347]: Same comment as above

File: src/core/solvers/ddp.cpp
Comment (andreadelprete) [reply src/core/solvers/ddp.cpp:360]: I've included only the Quu_inv*Qux because I'm currently profiling only the cubic-complexity operations. I can change the name as you suggested

**Comment by andreadelprete** [comment]
I should have addressed all your comments (thanks!) with the last 2 commits, except for the licensing thing, for which I wait for instructions.

File: include/crocoddyl/core/utils/Stdafx.hh
Comment (cmastalli) [reply include/crocoddyl/core/utils/Stdafx.hh:6]: I am fine with whatever action @nim65s thinks is needed here.

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [reply src/core/solvers/ddp.cpp:347]: We re-implement this function in `BoxDDP`. See this code: https://github.com/loco-3d/crocoddyl/blob/devel/src/core/solvers/box-ddp.cpp#L47-L78

It is the same for the `BoxFDDP` as you can see here: https://github.com/loco-3d/crocoddyl/blob/devel/src/core/solvers/box-fddp.cpp#L47-L78

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [reply src/core/solvers/ddp.cpp:128]: You're right for `computeDirection`, but not for the other points.

I will mark this conversation as resolved :)

File: src/core/solvers/ddp.cpp
Comment (cmastalli) [reply src/core/solvers/ddp.cpp:300]: The forward passed for `FDDP`, `BoxDDP`, `BoxFDDP` are different. We don't inherit them.

File: include/crocoddyl/core/utils/Stdafx.hh
Comment (nim65s) [reply include/crocoddyl/core/utils/Stdafx.hh:6]: @andreadelprete : I think you should try to contact Tommaso Urli to ask him if he is OK to let you use this code in a BSD licensed project.

**Comment by andreadelprete** [comment]
> Thanks @andreadelprete for this nice work.
> 
> The most relevant feedback is about propagating the profiling code to the missing pieces in FDDP, BoxDDP and BoxFDDP solvers. There are also other places that might be useful (e.g., shooting problem functions).

I see. However, the goal of this PR was simply to introduce the stop-watch class and to demonstrate its usage (in the DDP solver). I would add the code for profiling all the solvers in a future PR.

**Review by cmastalli — APPROVED** [review]
Thanks again @andreadelprete.
It is fine if we propagate later this profiling-related code to other solvers.

We could merge this PR after passing the jobs. Go ahead if there is nothing left!

**Comment by andreadelprete** [comment]
> Thanks again @andreadelprete.
> It is fine if we propagate later this profiling-related code to other solvers.
> 
> We could merge this PR after passing the jobs. Go ahead if there is nothing left!

Great, but I don't get why the jobs are not passing. 
One of them (972.3) is failing with this message
`The command "travis_wait 60 /bin/bash .travis/check-formatting.sh" exited with 122.`

The other one (972.2) is failing while installing the dependencies.

Finally, if by "go ahead" you meant "merge the PR", then I cannot do that because I don't have writing rights on this repo.

**Comment by cmastalli** [comment]
Hi @andreadelprete -- I have pushed a couple commits that fix the code format. The only pending discussion is about the BSD license request. If he accepts, then we could write the copyright note as other files (with of course his name on the file).

On the other hand, you can merge this PR, as I approved, if the CI jobs are passing.

**Comment by cmastalli** [comment]
Hi @andreadelprete -- I have pushed a couple commits that fix the code format. The only pending discussion is about the BSD license request. If he accepts, then we could write the copyright note as other files (with of course his name on the file).

On the other hand, you can merge this PR, as I approved, if the CI jobs are passing.

**Comment by andreadelprete** [comment]
> Hi @andreadelprete -- I have pushed a couple commits that fix the code format. The only pending discussion is about the BSD license request. If he accepts, then we could write the copyright note as other files (with of course his name on the file).

I've asked for permission to use the BSD license. I've let you know as soon as I get an answer.
 
> On the other hand, you can merge this PR, as I approved, if the CI jobs are passing.

I don't have writing rights on this repo. If you give them to me I will take care of the merge.

**Comment by cmastalli** [comment]
@andreadelprete I will merge this as I don't think it is useful to wait for author answer.

If he accepts, then we could create another PR.

Again, thanks for this nice work :)

**Pull request merged**
